### DISCUSSION TITLE
Discourage impl of trait not in your crate for type not in your crate

### DISCUSSION THREAD

**Discussion by dtolnay** [rust-lang-nursery/api-guidelines#138] [API Guidelines]
Such impls can lead to surprising and hard to debug inference failures in totally unrelated code. For example:

```rust
fn main() {
    let bools = vec![true, false];
    let stuff = Vec::new();
    assert_ne!(bools, stuff);
}
```

But adding this impl makes the inference break. The impl is PartialEq (not in our crate) for bool (not in our crate).

```rust
struct S;
impl PartialEq<S> for bool {
    fn eq(&self, other: &S) -> bool {
        unimplemented!()
    }
}
```

Real instance of this causing trouble: https://github.com/diesel-rs/diesel/commit/1f3d7b1d521d616ebd969a5c19c375807579d864

**Comment by Ixrec** [comment]
This seems like it's harmful primarily when the trait and type are both from `std`. Maybe we could add a clippy lint against the general pattern of `impl<...> [std trait]<...> for [std type]<...>`?

**Reply by dtolnay** [reply]
I don't see why std would be any different than things from any other crate when it comes to this behavior.
